### ISSUE TITLE
TabNet Training is slow

### ISSUE THREAD

**Issue by ptuls** [google-research/google-research#243]
First of, I'd like to thank the authors for their interesting work! I'm not sure how to make a pull request to this repository, so I'll leave my thoughts here regarding TabNet training. 

**Problem**:
TabNet training is slow because the parsing of the CSV is performed on the entire file each time, instead of performed over a single batch each time (see [line](https://github.com/google-research/google-research/blob/432ad195830f3722e2d2bc6dfe5f39991c384806/tabnet/data_helper_covertype.py#L126)). This results in overhead in CPU processing, reducing the utilization of the GPU. 

**Solution**:
In order to improve the training speed, we set the batching operation before the mapping operation so that the mapping operation only operates on a smaller set of data each iteration as follows:
```
    dataset = dataset.batch(batch_size, drop_remainder=True)
    dataset = dataset.map(parse_csv, num_parallel_calls=n_parallel)

    # Repeat after shuffling, to prevent separate epochs from blending together.
    dataset = dataset.repeat(num_epochs)
    return dataset
```
In terms of GPU utilization on an RTX2080Ti, I am getting about 70% utilization compared to 28% previously. More could potentially be done but this is a start.

**Caveat**:
The unfortunate side-effect of this is that the last small batch would be dropped. Fortunately this has not impacted accuracy on a test set.

Padding was tried, but I think I am not getting the shapes right, so I've used this as a solution in the interim.

**Testing**:
Shown here are the test accuracy and test loss when tested on the CoverType dataset. When tested on the CoverType dataset, the accuracy reached on the test set is 96.23% for the final model.

![image](https://user-images.githubusercontent.com/1995167/79059722-bcb89580-7cc0-11ea-9bb6-3087f2a58fdf.png)

![image](https://user-images.githubusercontent.com/1995167/79059768-2f297580-7cc1-11ea-9027-2acf825d0eb3.png)

**Comment by ptuls** [comment]
Closing, made pull request #244

**Comment by yhkll** [comment]
> > First of, I'd like to thank the authors for their interesting work! I'm not sure how to make a pull request to this repository, so I'll leave my thoughts here regarding TabNet training.
> > **Problem**: TabNet training is slow because the parsing of the CSV is performed on the entire file each time, instead of performed over a single batch each time (see [line](https://github.com/google-research/google-research/blob/432ad195830f3722e2d2bc6dfe5f39991c384806/tabnet/data_helper_covertype.py#L126)). This results in overhead in CPU processing, reducing the utilization of the GPU.
> > **Solution**: In order to improve the training speed, we set the batching operation before the mapping operation so that the mapping operation only operates on a smaller set of data each iteration as follows:
> > ```
> >     dataset = dataset.batch(batch_size, drop_remainder=True)
> >     dataset = dataset.map(parse_csv, num_parallel_calls=n_parallel)
> > 
> >     # Repeat after shuffling, to prevent separate epochs from blending together.
> >     dataset = dataset.repeat(num_epochs)
> >     return dataset
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In terms of GPU utilization on an RTX2080Ti, I am getting about 70% utilization compared to 28% previously. More could potentially be done but this is a start.
> > **Caveat**: The unfortunate side-effect of this is that the last small batch would be dropped. Fortunately this has not impacted accuracy on a test set.
> > Padding was tried, but I think I am not getting the shapes right, so I've used this as a solution in the interim.
> > **Testing**: Shown here are the test accuracy and test loss when tested on the CoverType dataset. When tested on the CoverType dataset, the accuracy reached on the test set is 96.23% for the final model.
> > ![image](https://user-images.githubusercontent.com/1995167/79059722-bcb89580-7cc0-11ea-9bb6-3087f2a58fdf.png)
> > ![image](https://user-images.githubusercontent.com/1995167/79059768-2f297580-7cc1-11ea-9027-2acf825d0eb3.png)
> 
> HI! I met the same problem. The TabNet Training is slow coping with [38*130000] data. Did you process the raw data or do sth. special to increase the speed? How you get it? Highly appreciated for any information and suggestions

Hello,as the README.md said,I use `python -m test_experiment_covertype.py` for simple test,but its result is obviously wrong,The traing loss and val accuracy are irregular,just because it's  in the low-resource environment?How can I fix this problem?
![KH HVLC(7N@1KB$A@Y%FBDF](https://user-images.githubusercontent.com/41032255/139565137-ac318333-3c6e-4833-84a3-33320e4357b0.png)
